### PR TITLE
Use C++1x variardic templates for generating redirections

### PR DESCRIPTION
**Pull request by 3v1n0**
This is based on some other my branches, so it's not yet ready to merge, but mostly here for a discussion.

Basically I've rewritten the parts to generate function wrappers using C++14 variardic templates which allows to generate code at compile time using static checks... Basically everything can be done with just one template-function `redirect_n` which generates all the needed functions without repeating the same code with macros, which is error-prone and not easily maintenable (as code errors can't be easily tracked).

I've also added a couple of missing syscalls (`xstat`, `scandir64` and `scandirat64`) and using static variables to hold the function pointers, to avoid calling `dlsym` multiple times on the same names.

No further optimizations for now (although I've another branch for some string changes), as I'd like to discuss about this approach, as I find it to be cleaner than leaving the macros all that power.

### CODE REVIEW

**Comment by 3v1n0** [comment]
@kyrofa

**Comment by sergiusens** [comment]
if this goes the c++ route can we also use `std::string` and its comparison methods and expose it using `.c_string` as well?

**Comment by sergiusens** [comment]
Also, conflict

**Comment by 3v1n0** [comment]
> if this goes the c++ route can we also use std::string and its comparison methods and expose it using .c_string as well?

Sure, and it's already mostly done in this branch:
https://github.com/3v1n0/snapcraft-preload/commit/0a1d2c029b371f93389b63e5e5f685afded462af

I would keep the two things in different PRs, though not to make reviews harder.
It's even true that using C-strings more smartly would allow to avoid copies in the many cases in which the redirected path is the same of the original one.

> Also, conflict

Yeah, that was planned as said in the first message, since it was based on part of other my branches (which I'd like to land before of this anyway).

**Comment by 3v1n0** [comment]
I've updated this with some more cleanups, would you mind to check this?

I was thinking also about using [boost preprocessor library](http://www.boost.org/doc/libs/1_63_0/libs/preprocessor/doc/index.html) to generate the calls better, but that's probably not needed as what we've now is enough generic for what we need.

File: preload.cpp [hunk 1]
```diff
@@ -54,6 +59,14 @@ static char *saved_snap_revision = NULL;
 static size_t saved_snap_revision_len = 0;
 static char saved_snap_devshm[NAME_MAX];
 
+static int (*_access) (const char *, int) = NULL;
```
Comment (kyrofa) [inline preload.cpp]: Due to the anonymous namespace all of these already have internal linkage, so we can probably drop the `static`s on these, no?

File: preload.cpp [hunk 2]
```diff
@@ -54,6 +59,14 @@ static char *saved_snap_revision = NULL;
...
+
+template <typename dirent_t>
+using filter_function_t = int (*)(const dirent_t *);
+template <typename dirent_t>
+using compar_function_t = int (*)(const dirent_t **, const dirent_t **);
```
Comment (kyrofa) [inline preload.cpp]: Should this be `compare_function_t`?

File: preload.cpp [hunk 3]
```diff
@@ -130,7 +145,7 @@ redirect_writable_path (const char *pathname, const char *basepath)
         return strdup (basepath);
     }
 
-    redirected_pathname = malloc (PATH_MAX);
+    redirected_pathname = static_cast<char *> (malloc (PATH_MAX));
```
Comment (kyrofa) [inline preload.cpp]: Blech. Is this something you're fixing later, then?

File: preload.cpp [hunk 4]
```diff
@@ -227,196 +241,152 @@ redirect_path_full (const char *pathname, int check_parent, int only_if_absolute
...
-    RET result; \
-    _NAME = (RET (*)(const char *path)) dlsym (RTLD_NEXT, #NAME); \
-    new_path = redirect_path (path); \
-    result = _NAME (new_path); \
-    free (new_path); \
-    return result; \
+// helper class
+template<typename R, template<typename...> class Params, typename... Args, std::size_t... I>
+inline R call_helper(std::function<R(Args...)> const&func, Params<Args...> const&params, std::index_sequence<I...>)
+{ return func(std::get<I>(params)...); }
```
Comment (kyrofa) [inline preload.cpp]: Personally not a fan of the formatting here (I'd keep it looking like the rest).

File: preload.cpp (hunk 2 above)
Comment (3v1n0) [inline preload.cpp]: It's called `compar` in the default C definition:

```C
int scandir(const char *dirp, struct dirent ***namelist,
              int (*filter)(const struct dirent *),
              int (*compar)(const struct dirent **, const struct dirent **));
```

Thus I've kept the name, let me know if you still prefer it to be `compare`.

File: preload.cpp (hunk 3 above)
Comment (3v1n0) [inline preload.cpp]: Yeah, I've the branch for using `std::string` instead, so, I'll change it.

But a `static_cast` is still better than an implicit one ;-).

File: preload.cpp (hunk 4 above)
Comment (3v1n0) [inline preload.cpp]: Ack

File: preload.cpp (hunk 2 above)
Comment (kyrofa) [inline preload.cpp]: No that's fine.

**Comment by 3v1n0** [comment]
@kyrofa ok, this should be good for merging again.

**Comment by 3v1n0** [comment]
@sergiusens, @kyrofa any news on this?

**Comment by kyrofa** [comment]
Yeah, my +1 is still valid.

**Pull request merged**